File: packages/less/src/less/tree/js-eval-node.test.js
import { describe, it, expect } from 'vitest';
import JsEvalNode from './js-eval-node';
import Node from './node';

describe('JsEvalNode', () => {
    let jsEvalNode;
    let mockContext;

    beforeEach(() => {
        jsEvalNode = new JsEvalNode();
        // Setup basic mock context
        mockContext = {
            javascriptEnabled: true,
            frames: [
                {
                    // Implement both variable() and variables() methods
                    variable: (name) => {
                        const vars = {
                            '@color': {
                                value: {
                                    eval: () => ({
                                        toCSS: () => '"#ff0000"' // Quoted for JavaScript string interpolation
                                    })
                                }
                            },
                            '@size': {
                                value: {
                                    eval: () => ({
                                        toCSS: () => '"16px"' // Quoted for JavaScript string interpolation
                                    })
                                }
                            }
                        };
                        return vars[name];
                    },
                    variables: () => ({
                        '@color': {
                            value: {
                                eval: () => ({
                                    toCSS: () => '#ff0000' // Unquoted for toJS() calls
                                })
                            }
                        },
                        '@size': {
                            value: {
                                eval: () => ({
                                    toCSS: () => '16px' // Unquoted for toJS() calls
                                })
                            }
                        }
                    })
                }
            ],
            importantScope: [{ important: null }]
        };
    });

    describe('evaluateJavaScript', () => {
        it('throws error when JavaScript is not enabled', () => {
            mockContext.javascriptEnabled = false;
            expect(() => {
                jsEvalNode.evaluateJavaScript('1 + 1', mockContext);
            }).toThrow('Inline JavaScript is not enabled');
        });

        it('evaluates basic JavaScript expressions', () => {
            const result = jsEvalNode.evaluateJavaScript('1 + 1', mockContext);
            expect(result).toBe(2);
        });

        it('interpolates Less variables in JavaScript expressions', () => {
            jsEvalNode._fileInfo = { filename: 'test.less' };
            const result = jsEvalNode.evaluateJavaScript(
                '"The color is " + @{color}',
                mockContext
            );
            expect(result).toBe('The color is #ff0000');
        });

        it('provides access to Less variables in evaluation context', () => {
            const result = jsEvalNode.evaluateJavaScript(
                'this.color.toJS()',
                mockContext
            );
            expect(result).toBe('#ff0000');
        });

        it('throws error on invalid JavaScript syntax', () => {
            jsEvalNode._fileInfo = { filename: 'test.less' };
            expect(() => {
                jsEvalNode.evaluateJavaScript('invalid syntax )', mockContext);
            }).toThrow('JavaScript evaluation error');
        });

        it('throws error on runtime JavaScript errors', () => {
            jsEvalNode._fileInfo = { filename: 'test.less' };
            expect(() => {
                jsEvalNode.evaluateJavaScript(
                    'undefinedFunction()',
                    mockContext
                );
            }).toThrow('JavaScript evaluation error');
        });

        it('handles multiple variable interpolations', () => {
            jsEvalNode._fileInfo = { filename: 'test.less' };
            const result = jsEvalNode.evaluateJavaScript(
                '"Size: " + @{size} + ", Color: " + @{color}',
                mockContext
            );
            expect(result).toBe('Size: 16px, Color: #ff0000');
        });

        it('handles variable interpolation with expressions', () => {
            mockContext.frames[0].variable = (name) =>
                ({
                    '@prefix': {
                        value: { eval: () => ({ toCSS: () => '"my-"' }) }
                    },
                    '@suffix': {
                        value: { eval: () => ({ toCSS: () => '"value"' }) }
                    }
                }[name]);

            const result = jsEvalNode.evaluateJavaScript(
                '@{prefix} + @{suffix}',
                mockContext
            );
            expect(result).toBe('my-value');
        });

        it('handles empty expressions', () => {
            const result = jsEvalNode.evaluateJavaScript('""', mockContext);
            expect(result).toBe('');
        });

        it('handles expressions returning complex objects', () => {
            const result = jsEvalNode.evaluateJavaScript(
                '({ key: "value" })',
                mockContext
            );
            expect(result).toEqual({ key: 'value' });
        });

        it('handles multiple variable access in same context', () => {
            const result = jsEvalNode.evaluateJavaScript(
                'this.color.toJS() + this.size.toJS()',
                mockContext
            );
            expect(result).toBe('#ff000016px');
        });

        it('handles special characters and escape sequences', () => {
            const result = jsEvalNode.evaluateJavaScript(
                '"\\n" + @{color} + "\\t"',
                mockContext
            );
            expect(result).toBe('\n#ff0000\t');
        });

        it('handles undefined variables in context', () => {
            mockContext.frames[0].variables = () => ({});
            const result = jsEvalNode.evaluateJavaScript(
                'this.undefinedVar',
                mockContext
            );
            expect(result).toBe(undefined);
        });

        it('throws on invalid variable interpolation', () => {
            expect(() => {
                jsEvalNode.evaluateJavaScript('@{nonexistent}', mockContext);
            }).toThrow();
        });

        it('handles special characters in variable names', () => {
            mockContext.frames[0].variable = (name) =>
                ({
                    '@special-name_123': {
                        value: { eval: () => ({ toCSS: () => '"special"' }) }
                    }
                }[name]);

            const result = jsEvalNode.evaluateJavaScript(
                '@{special-name_123}',
                mockContext
            );
            expect(result).toBe('special');
        });

        it('handles missing file info gracefully', () => {
            jsEvalNode._fileInfo = null;
            expect(() => {
                jsEvalNode.evaluateJavaScript('throw new Error()', mockContext);
            }).toThrow('JavaScript evaluation error');
        });

        it('handles numeric and boolean variable values', () => {
            mockContext.frames[0].variable = (name) =>
                ({
                    '@number': {
                        value: { eval: () => ({ toCSS: () => '42' }) }
                    },
                    '@bool': {
                        value: { eval: () => ({ toCSS: () => 'true' }) }
                    }
                }[name]);

            const numResult = jsEvalNode.evaluateJavaScript(
                '@{number} + 1',
                mockContext
            );
            expect(numResult).toBe(43);

            const boolResult = jsEvalNode.evaluateJavaScript(
                '@{bool}',
                mockContext
            );
            expect(boolResult).toBe(true);
        });

        it('handles multiple frames in context', () => {
            mockContext.frames.unshift({
                variables: () => ({
                    '@override': {
                        value: {
                            eval: () => ({
                                toCSS: () => 'frame0'
                            })
                        }
                    }
                })
            });

            const result = jsEvalNode.evaluateJavaScript(
                'this.override.toJS()',
                mockContext
            );
            expect(result).toBe('frame0');
        });

        it('handles complex objects in variable evaluation', () => {
            const complexObj = { foo: 'bar', arr: [1, 2] };
            mockContext.frames[0].variable = (name) =>
                ({
                    '@complex': {
                        value: {
                            eval: () => ({
                                toCSS: () => `(${JSON.stringify(complexObj)})`
                            })
                        }
                    }
                }[name]);

            const result = jsEvalNode.evaluateJavaScript(
                '@{complex}',
                mockContext
            );
            expect(result).toEqual(complexObj);
        });
    });

    describe('jsify', () => {
        it('converts single value to CSS string', () => {
            const obj = {
                value: 'test',
                toCSS: () => 'test'
            };
            expect(jsEvalNode.jsify(obj)).toBe('test');
        });

        it('converts array values to CSS array string', () => {
            const obj = {
                value: [{ toCSS: () => 'red' }, { toCSS: () => 'blue' }]
            };
            expect(jsEvalNode.jsify(obj)).toBe('[red, blue]');
        });

        it('handles single-item arrays', () => {
            const obj = {
                value: [{ toCSS: () => 'red' }],
                toCSS: () => 'red'
            };
            expect(jsEvalNode.jsify(obj)).toBe('red');
        });

        it('handles empty arrays', () => {
            const obj = {
                value: [],
                toCSS: () => ''
            };
            expect(jsEvalNode.jsify(obj)).toBe('');
        });

        it('handles nested arrays', () => {
            const obj = {
                value: [{ toCSS: () => '[1, 2]' }, { toCSS: () => '[3, 4]' }]
            };
            expect(jsEvalNode.jsify(obj)).toBe('[[1, 2], [3, 4]]');
        });

        it('handles objects without toCSS method', () => {
            const obj = {
                value: [{}]
            };
            expect(() => jsEvalNode.jsify(obj)).toThrow();
        });

        it('handles null values', () => {
            const obj = {
                value: null,
                toCSS: () => 'null'
            };
            expect(jsEvalNode.jsify(obj)).toBe('null');
        });

        it('handles malformed value property', () => {
            const obj = {
                toCSS: () => 'default'
            };
            expect(jsEvalNode.jsify(obj)).toBe('default');
        });

        it('handles values with special characters', () => {
            const obj = {
                value: [
                    { toCSS: () => 'test\nline' },
                    { toCSS: () => 'test\t tab' }
                ]
            };
            expect(jsEvalNode.jsify(obj)).toBe('[test\nline, test\t tab]');
        });

        it('handles mixed types in arrays', () => {
            const obj = {
                value: [
                    { toCSS: () => '42' },
                    { toCSS: () => '"string"' },
                    { toCSS: () => 'true' }
                ]
            };
            expect(jsEvalNode.jsify(obj)).toBe('[42, "string", true]');
        });

        it('handles errors in toCSS calls', () => {
            const obj = {
                value: [
                    {
                        toCSS: () => {
                            throw new Error('toCSS failed');
                        }
                    }
                ],
                toCSS: () => {
                    throw new Error('toCSS failed');
                }
            };
            expect(() => jsEvalNode.jsify(obj)).toThrow(/toCSS failed/);
        });

        it('handles large arrays', () => {
            const largeArray = Array(1000).fill({ toCSS: () => '1' });
            const obj = { value: largeArray };
            const result = jsEvalNode.jsify(obj);
            expect(result.split(',').length).toBe(1000);
        });
    });

    describe('integration with Node prototype', () => {
        it('inherits from Node', () => {
            expect(jsEvalNode).toBeInstanceOf(Node);
        });

        it('maintains file info and index from parent Node', () => {
            jsEvalNode._fileInfo = { filename: 'test.less' };
            jsEvalNode._index = 1;

            expect(jsEvalNode.fileInfo()).toEqual({ filename: 'test.less' });
            expect(jsEvalNode.getIndex()).toBe(1);
        });
    });
});

File: packages/less/src/less/less_go/js_eval_node_test.go
package less_go

import (
	"strings"
	"testing"
)

// jsTestFrame implements Frame interface for testing
type jsTestFrame struct {
	vars map[string]map[string]any
}

func (m *jsTestFrame) Variable(name string) map[string]any {
	return m.vars[name]
}

func (m *jsTestFrame) Property(name string) []any {
	// Return empty array for now - tests don't use properties
	return nil
}

func (m *jsTestFrame) Variables() map[string]any {
	// Convert map[string]map[string]any to map[string]any
	result := make(map[string]any)
	for k, v := range m.vars {
		result[k] = v
	}
	return result
}

// jsTestContext implements a minimal context for testing
type jsTestContext struct {
	javascriptEnabled bool
	frames            []ParserFrame
	importantScope    []map[string]bool
	inCalc            bool
	mathOn            bool
}

func (m *jsTestContext) IsJavaScriptEnabled() bool {
	return m.javascriptEnabled
}

func (m *jsTestContext) GetFrames() []ParserFrame {
	return m.frames
}

func (m *jsTestContext) GetImportantScope() []map[string]bool {
	return m.importantScope
}

func (m *jsTestContext) IsInCalc() bool {
	return m.inCalc
}

func (m *jsTestContext) IsMathOn() bool {
	return m.mathOn
}

func (m *jsTestContext) SetMathOn(on bool) {
	m.mathOn = on
}

func (m *jsTestContext) EnterCalc() {
	m.inCalc = true
}

func (m *jsTestContext) ExitCalc() {
	m.inCalc = false
}

func (m *jsTestContext) GetDefaultFunc() *DefaultFunc {
	return nil
}

// jsTestEvalValue implements the Eval method returning both a value and error
type jsTestEvalValue struct {
	value any
	err   error
}

func (m *jsTestEvalValue) Eval(ctx EvalContext) (any, error) {
	return m.value, m.err
}

// jsTestCSSable is a mock implementation for values with ToCSS
type jsTestCSSable struct {
	css string
}

func (m *jsTestCSSable) ToCSS() string {
	return m.css
}

func TestJsEvalNode(t *testing.T) {
	var jsEvalNode *JsEvalNode
	var mockCtx *jsTestContext

	setup := func() {
		jsEvalNode = NewJsEvalNode()
		
		// Create mock values for variables
		colorValue := &jsTestEvalValue{
			value: map[string]any{
				"value": &jsTestCSSable{css: "#ff0000"},
			},
		}
		
		sizeValue := &jsTestEvalValue{
			value: map[string]any{
				"value": &jsTestCSSable{css: "16px"},
			},
		}
		
		// Create a basic mock context
		mockCtx = &jsTestContext{
			javascriptEnabled: true,
			frames: []ParserFrame{
				&jsTestFrame{
					vars: map[string]map[string]any{
						"@color": {
							"value": colorValue,
						},
						"@size": {
							"value": sizeValue,
						},
					},
				},
			},
			importantScope: []map[string]bool{
				{"important": false},
			},
		}
	}

	t.Run("evaluateJavaScript", func(t *testing.T) {
		t.Run("throws error when JavaScript is not enabled", func(t *testing.T) {
			setup()
			mockCtx.javascriptEnabled = false
			jsEvalNode.SetFileInfo(map[string]any{"filename": "test.less"})
			
			_, err := jsEvalNode.EvaluateJavaScript("1 + 1", mockCtx)
			if err == nil {
				t.Error("Expected error when JavaScript is not enabled")
			}
			if !strings.Contains(err.Error(), "inline JavaScript is not enabled") {
				t.Errorf("Unexpected error message: %s", err.Error())
			}
		})

		t.Run("returns not supported error when JavaScript is enabled", func(t *testing.T) {
			setup()
			jsEvalNode.SetFileInfo(map[string]any{"filename": "test.less"})
			
			_, err := jsEvalNode.EvaluateJavaScript("1 + 1", mockCtx)
			if err == nil {
				t.Error("Expected error indicating JS evaluation is not supported")
			}
			if !strings.Contains(err.Error(), "JavaScript evaluation is not supported") {
				t.Errorf("Expected error about JS evaluation not supported, got: %s", err.Error())
			}
		})

		t.Run("interpolates variables in error message", func(t *testing.T) {
			setup()
			jsEvalNode.SetFileInfo(map[string]any{"filename": "test.less"})
			
			_, err := jsEvalNode.EvaluateJavaScript("@{color} + @{size}", mockCtx)
			if err == nil {
				t.Error("Expected error indicating JS evaluation is not supported")
			}
			errMsg := err.Error()
			if !strings.Contains(errMsg, "#ff0000") || !strings.Contains(errMsg, "16px") {
				t.Errorf("Expected error message to contain interpolated values, got: %s", errMsg)
			}
		})
	})

	t.Run("jsify", func(t *testing.T) {
		t.Run("converts single value to CSS string", func(t *testing.T) {
			setup()
			obj := &jsTestCSSable{css: "test"}
			result := jsEvalNode.jsify(obj)
			if result != "test" {
				t.Errorf("Expected 'test', got '%s'", result)
			}
		})

		t.Run("converts array values to CSS array string", func(t *testing.T) {
			setup()
			obj := []any{
				&jsTestCSSable{css: "red"},
				&jsTestCSSable{css: "blue"},
			}
			result := jsEvalNode.jsify(obj)
			if result != "[red, blue]" {
				t.Errorf("Expected '[red, blue]', got '%s'", result)
			}
		})

		t.Run("handles nil values", func(t *testing.T) {
			setup()
			result := jsEvalNode.jsify(nil)
			if result != "null" {
				t.Errorf("Expected 'null', got '%s'", result)
			}
		})

		t.Run("extracts value from result object", func(t *testing.T) {
			setup()
			obj := map[string]any{
				"value": &jsTestCSSable{css: "extracted"},
			}
			result := jsEvalNode.jsify(obj)
			if result != "extracted" {
				t.Errorf("Expected 'extracted', got '%s'", result)
			}
		})
	})

	t.Run("node integration", func(t *testing.T) {
		t.Run("inherits from Node", func(t *testing.T) {
			jsEvalNode := NewJsEvalNode()
			if jsEvalNode.Node == nil {
				t.Error("Expected JsEvalNode to inherit from Node")
			}
		})

		t.Run("has correct type", func(t *testing.T) {
			jsEvalNode := NewJsEvalNode()
			if jsEvalNode.GetType() != "JsEvalNode" {
				t.Errorf("Expected type 'JsEvalNode', got '%s'", jsEvalNode.GetType())
			}
		})
	})
} 